pkg/clients/resourcerecordset/resourcerecordset.go

```diff
@@ -0,0 +1,177 @@
+type NotFound struct{}
+
+func (err *NotFound) Error() string {
+	return fmt.Sprint(RRSetNotFound)
+}
+
+// NewClient creates new AWS client with provided AWS Configuration/Credentials
+func NewClient(config *aws.Config) Client {
+	return route53.New(*config)
+}
+
+// GenerateChangeResourceRecordSetsInput prepares input for a ChangeResourceRecordSetsInput
+func GenerateChangeResourceRecordSetsInput(p *v1alpha3.ResourceRecordSetParameters, action route53.ChangeAction) *route53.ChangeResourceRecordSetsInput {
+	var ttl *int64
+	if p.TTL == nil {
+		num := int64(300)
+		ttl = &num
+	} else {
+		ttl = p.TTL
+	}
+
+	resourceRecords := make([]route53.ResourceRecord, 0)
+	for _, r := range p.Records {
+		record := r
+		resourceRecords = append(resourceRecords, route53.ResourceRecord{
+			Value: &record,
+		})
+	}
+
+	resourceRecordSet := route53.ResourceRecordSet{
+		Name:            p.Name,
+		Type:            route53.RRType(aws.StringValue(p.Type)),
+		TTL:             ttl,
+		ResourceRecords: resourceRecords,
+	}
+	changeBatch := &route53.ChangeBatch{
+		Changes: []route53.Change{
+			{
+				Action:            action,
+				ResourceRecordSet: &resourceRecordSet,
+			},
+		},
+	}
+	return &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: p.ZoneID,
+		ChangeBatch:  changeBatch,
+	}
+}
+
+// GenerateObservation is used to produce v1alpha3.ResourceRecordSetObservation from
+// route53.ResourceRecordSet
+func GenerateObservation(r route53.ResourceRecordSet) v1alpha3.ResourceRecordSetObservation {
+
+	rType := string(r.Type)
+	resourceRecords := make([]*v1alpha3.ResourceRecord, 0)
+	for _, item := range r.ResourceRecords {
+		resourceRecords = append(resourceRecords, &v1alpha3.ResourceRecord{Value: item.Value})
+	}
+
+	o := v1alpha3.ResourceRecordSetObservation{
+		Name:            r.Name,
+		Type:            &rType,
+		TTL:             r.TTL,
+		ResourceRecords: resourceRecords,
+	}
+
+	return o
+}
+
+// IsUpToDate checks if object is up to date
+func IsUpToDate(p v1alpha3.ResourceRecordSetParameters, rrset route53.ResourceRecordSet) (bool, error) {
+	patch, err := CreatePatch(&rrset, &p)
+	if err != nil {
+		return false, err
+	}
+	return cmp.Equal(&v1alpha3.ResourceRecordSetParameters{}, patch), nil
+}
+
+// LateInitialize fills the empty fields in *v1alpha3.ResourceRecordSetParameters with
+// the values seen in route53.ResourceRecordSet.
+func LateInitialize(in *v1alpha3.ResourceRecordSetParameters, rrSet *route53.ResourceRecordSet) {
+	if rrSet == nil {
+		return
+	}
+
+	in.Name = awsclients.LateInitializeStringPtr(in.Name, rrSet.Name)
+
+	rrType := string(rrSet.Type)
+	in.Type = awsclients.LateInitializeStringPtr(in.Type, &rrType)
+
+	in.TTL = awsclients.LateInitializeInt64Ptr(in.TTL, rrSet.TTL)
+	if len(in.Records) == 0 && len(rrSet.ResourceRecords) != 0 {
+		in.Records = make([]string, len(rrSet.ResourceRecords))
+		for i, val := range rrSet.ResourceRecords {
+			in.Records[i] = aws.StringValue(val.Value)
+		}
+	}
```

Comment:
Do we need to handle the case in which we have a partial view of the records in our spec? i.e. Perhaps `spec.records` has two items, but there are actually three items according to the AWS API?

Reply:
For the resource records, we are maintaining our Object spec as the final source of truth. Therefore in such a case, I believe the additional record will be updated with what's available in the spec.